### PR TITLE
Fix out-of-bounds writing to browser messages

### PR DESCRIPTION
**Pull request by varjolintu**
[NOTE]: # ( Describe your changes in detail. Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX". )
[NOTE]: # ( If you used Generative AI to write the majority of your code, you must state this. )
Adds safeguards when encrypting or decrypting messages for browser integration / native messaging.

Socket/pipes are transferring base64 encoded messages that are always larger than the actual message. The socket buffer must remain with 1MB. Raw data in the decrypted message is reduced to 655 360 bytes, so there is enough space for the base64 encoding.

Both encrypt and decrypt now checks if the message, public key and nonce lengths are valid (when incrementing a nonce as well). For encryption, the size of the buffer must be size+16 (this is what `crypto_box_easy()` requires). For decryption, the base64 decoded buffer holds that extra 16 bytes (macBytes), and the buffer for decrypted data is added with that amount.

### Support for fragmented and coalesced messages
BrowserHost class now supports fragmented and coalesced messages. When a message arrives to the local socket, it is checked if it's coalesced (meaning the message buffer holds multiple JSON objects). After that every message is checked for fragmentation (message arrives in multiple socket messages). Combined fragmented messages cannot exceed the maximum buffer size. Native messaging supports receiving only full JSON objects.

This feature relies on the browser integration where messages are always a single JSON object with only string values.

If a message is detected as fragmented, the following steps are made:
1. The first message part is stored to a temporary map, identified by the current socket descriptor.
2. Messages after that are combined with the previous messages in the temporary map, and parsed again.
3. If a complete message is parsed correctly, it is returned and will be sent forward.
4. If a complete message size exceeds the maximum socket buffer size, it is discarded with a warning.

If a message has some garbage at the start or at the end, it is removed before parsing.

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and include helpful comments. )
Automatic tests added for situations where key and nonce lengths are not valid. Modified the encrypt/decrypt functions to take the max length as a parameter, so proper tests can be constructed. Tested the limits by reducing the max length of the message.

BrowserMessageBuilder (both encryption and decryption):
- Invalid public key, secret key or nonce.
- Max message length exceeded.
- Testing boundaries with maximum message length.
- Testing boundaries with macBytes.
- Testing boundaries with both maximum message length and macBytes.
- Null characters inside JSON.
- Malformed base64.
- Increment nonce checks the nonce length.

BrowserHost:
- Normal and non-valid JSON messages.
- Garbaged messages, with and without fragmentation.
- Different kind of fragmented messages.
- Maximum combined message size limit checks.
- Fragmented messages from two different sockets at the same time.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change that fixes an issue)

### CODE REVIEW

**Comment by sigure21** [comment]
## Reporter retest / review notes

I reviewed the current patch against the exact boundary cases from my original report.

### Environment and scope

**Original reproduction environment (reference):**
- Windows 11 Pro 10.0.26200, 64-bit
- Official KeePassXC 2.7.12 Windows 64-bit ZIP
- Microsoft CDB 10.0.29617.1000
- Isolated configuration and synthetic KDBX files; no personal database
- Authenticated browser-protocol client connected directly to KeePassXC's per-user named pipe

**Patch review environment:**
- PR head: `67c499cc86712cf394cb5059da6041cc26cd1f9c`
- Base: `84658ade9a1763dae6108d489a88e503ce3d146d`
- Source-path review of `buildResponse() -> encryptMessage() -> encrypt()`
- Exact byte-size recalculation using the plaintext/ciphertext lengths captured in the original tests
- Current GitHub/TeamCity check results reviewed

This comment distinguishes source/CI verification from runtime verification: I did not mark the patched binary as runtime-tested because the current head does not compile in several required CI configurations.

### Procedure

1. Confirmed that `encrypt()` converts the response to UTF-8 and rejects `messageBytes.length() > 1,048,576` before calling `crypto_box_easy()`.
2. Confirmed that accepted inputs allocate `maxLength + crypto_box_MACBYTES`, so the output buffer is large enough for every plaintext that reaches the encryption call.
3. Applied the exact plaintext/ciphertext sizes from the original small-control and boundary cases to the new condition.
4. Checked the failure path: an oversized response causes `encryptMessage()` to return empty and `buildResponse()` to return `ERROR_KEEPASS_CANNOT_ENCRYPT_MESSAGE` (error code 7).
5. Reviewed the added unit tests and the current TeamCity results.

### Boundary-case results

| Original scenario | Plaintext bytes | Ciphertext bytes | Original result | Result with this code path |
|---|---:|---:|---|---|
| 20 matching entries, title length 4,096 | 85,229 | 85,245 | Response completed; process survived | Accepted; output buffer is sufficient |
| 125 matching entries, title length 8,130 | 1,048,747 | 1,048,763 | 187-byte logical OOB; process survived | Rejected 171 bytes over plaintext limit; encryption is not called |
| 125 matching entries, title length 8,160 | 1,052,497 | 1,052,513 | 3,937-byte logical OOB; process survived | Rejected 3,921 bytes over plaintext limit; encryption is not called |
| 125 matching entries, title length 8,161 | 1,052,622 | 1,052,638 (calculated) | Process exited `0xC0000005` | Rejected 4,046 bytes over plaintext limit; encryption is not called |

### Result for the reported vulnerability

The reported `crypto_box_easy()` out-of-bounds write is **logically blocked by this patch**: all three oversized reporter cases return before the vulnerable call, while the small control remains below the limit. The allocation also covers the 16-byte MAC for every accepted plaintext.

I cannot yet call the fix fully verified or merge-ready for the following reasons:

1. **Current head is build-broken in required configurations.** TeamCity Debug Clang (292317), Debug GCC (292318), Debug No Features GCC (292319), and Coverage (292328) fail at `BrowserMessageBuilder.cpp:264` because `qsizetype` is compared with Botan's unsigned `crypto_box_MACBYTES` under `-Werror=sign-compare`. The previous commit passed these builds.
2. **The 1 MiB limit is applied to plaintext, not the final native message.** A maximum accepted 1,048,576-byte plaintext becomes a 1,048,592-byte ciphertext, a 1,398,124-byte Base64 value, and about a 1,398,203-byte compact `get-logins` outer JSON message. Chrome's host-to-browser native-message limit is 1 MiB, so the final serialized JSON needs a separate bound.
3. **The report's framing recommendation is not implemented.** Increasing `QLocalSocket` buffers to 2 MiB does not create message boundaries. Both directions still treat each `readAll()` chunk as a complete message, so fragmented or coalesced frames remain unhandled.
4. **Decrypt hardening is incomplete.** Please reject ciphertext shorter than `crypto_box_MACBYTES`, allocate exactly `ciphertext.size() - MAC`, return the explicit plaintext length instead of using `char_traits::length()`, bound encoded input before Base64 decoding, and validate the nonce before `incrementNonce()`.
5. **The tests do not cover the actual boundary shape.** Please add `limit-1/limit/limit+1`, `MAC-1/MAC/MAC+1`, malformed/oversized Base64, embedded NUL, fragmented/coalesced local-socket frames, and an ASan regression using the original oversized-response shape.

For the immediate signedness/bounds issue, a safe starting point is:

```cpp
constexpr auto MacBytes = static_cast<qsizetype>(crypto_box_MACBYTES);
if (maxLength < 0 || encryptedBytes.size() < MacBytes
    || encryptedBytes.size() - MacBytes > maxLength) {
    return {};
}
```

**Verdict:** the specific encryption OOB write I reported is addressed at the source-code level, but the current head needs the build fix and a patched-binary rerun of the original boundary scenarios before I would mark the remediation complete.

**Comment by varjolintu** [comment]
One note: This was difficult to test against a real proxy environment. I tried to decrease the buffer sizes to force normal messages to be fragmented, and sometimes on those cases the browser extension side complained that the JSON retrieved was not a real JSON data. And there were two packets incoming instead of one.

I can try to do some further testing where I force two separate packets to be seen as one coalesced.

**Comment by sigure21** [comment]
I checked the update.
I'll test it out.
Since today and tomorrow are Korean holidays, I need to go to my
grandmother at the hospital, so I can take the test starting in two days.
Would that be alright?

2026년 9월 21일 (월) 오후 1:52, Sami Vänttinen ***@***.***>님이 작성:

File: src/browser/BrowserHost.h [hunk 1]
```diff
@@ -15,10 +15,13 @@
  *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#ifndef KEEPASSXC_NATIVEMESSAGINGHOST_H
-#define KEEPASSXC_NATIVEMESSAGINGHOST_H
+#ifndef KEEPASSXC_BROWSERHOST_H
+#define KEEPASSXC_BROWSERHOST_H
```
Comment (droidmonkey) [inline src/browser/BrowserHost.h:19]: These should be `#pragma once`

File: src/browser/BrowserMessageBuilder.h [hunk 2]
```diff
@@ -18,10 +18,18 @@
 #ifndef KEEPASSXC_BROWSERMESSAGEBUILDER_H
 #define KEEPASSXC_BROWSERMESSAGEBUILDER_H
```
Comment (droidmonkey) [inline src/browser/BrowserMessageBuilder.h:19]: `#pragma once`

**Review by droidmonkey — APPROVED** [review]

**Comment by phoerious** [comment]
@sigure21 Merged this PR. We'll ship it in the beta. Please test anyway. If the fix isn't adequate, we can add further fixes in the final version.

**Comment by sigure21** [comment]
@phoerious @varjolintu I tested the merged fix as requested. The reported oversized browser-response encryption path is blocked in the source build I tested. I also investigated the real proxy fragmentation/coalescing behavior mentioned above; that produced a separate, reproducible JSON framing problem. Here are the exact methods, results, and limits of the tests.

### Tested code and environment

- Date: 2026-09-23 (KST).
- PR head: 3e23baddda5f1a5a5126dbe0d943698b6de839c9; merged commit: 80e68bb4e8b89e383dd96b1bf5de169ca61367aa; tested develop checkout: 9e0f57a4a4c6c629fa6d0a593acb7d089b1d95cd. The PR head and merged commit have the same stable git patch-id (dcc02da696ef3079a76c17e14b58443f680bfbca).
- Windows 11 host, WSL2 Ubuntu 24.04; GCC 13.3.0, Qt 6.4.2, Botan 2.19.3. Debug build with WITH_ASAN=ON and browser integration enabled.
- ASAN_OPTIONS=detect_leaks=0:halt_on_error=1. These runs checked memory accesses, not leaks.

### Build and browser regression suite

From the patched source checkout:

~~~sh
cmake -S . -B /root/kpxc-pr13679-build -G Ninja \
  -DCMAKE_BUILD_TYPE=Debug -DWITH_TESTS=ON -DWITH_GUI_TESTS=OFF \
  -DWITH_ASAN=ON -DWITH_X11=OFF -DKPXC_FEATURE_BROWSER=ON \
  -DKPXC_FEATURE_SSHAGENT=OFF -DKPXC_FEATURE_FDOSECRETS=OFF \
  -DKPXC_FEATURE_NETWORK=OFF -DKPXC_FEATURE_DOCS=OFF
ninja -C /root/kpxc-pr13679-build -j4 testbrowser keepassxc-proxy
cd /root/kpxc-pr13679-build/tests
QT_QPA_PLATFORM=offscreen ASAN_OPTIONS=detect_leaks=0:halt_on_error=1 ./testbrowser
~~~

Result: **41 passed, 0 failed, 0 skipped**, with no ASan finding. This includes invalid public/secret key and nonce lengths, encryption/decryption boundaries, malformed Base64, embedded NUL, and BrowserHost fragmented/coalesced message tests. Warnings from intentional invalid-input cases did not cause failures.

### Production-size encryption boundary

I separately compiled the merged src/browser/BrowserMessageBuilder.cpp with ASan and invoked its default production limit. The caller generated valid key pairs and a standard-Base64-encoded 24-byte nonce. It called encrypt() with plaintext lengths of exactly BrowserShared::NATIVEMSG_MAX_LENGTH (655,360), one byte above it (655,361), and 1,052,622 bytes (the plaintext length from my original Windows crash analysis). An oversized payload was also sent through buildResponse("get-logins", ...) to check the application-facing failure path. Core test logic:

~~~cpp
BrowserMessageBuilder builder;
const auto serverKeys = builder.getKeyPair();
const auto clientKeys = builder.getKeyPair();
const auto nonce = QString::fromLatin1(QByteArray(24, '\x01').toBase64());
const auto limit = BrowserShared::NATIVEMSG_MAX_LENGTH;

const auto atLimit = builder.encrypt(QString(limit, QLatin1Char('A')),
                                     nonce, clientKeys.first, serverKeys.second);
const auto overLimit = builder.encrypt(QString(limit + 1, QLatin1Char('B')),
                                       nonce, clientKeys.first, serverKeys.second);
const auto oldCrashSize = builder.encrypt(QString(1052622, QLatin1Char('C')),
                                          nonce, clientKeys.first, serverKeys.second);
// Also passed a 1,052,622-character payload via buildResponse("get-logins", ...).
~~~

Observed patched output:

~~~text
plaintext_limit=655360
exact_limit_ciphertext_base64_length=873836
one_past_limit_rejected=1
original_crash_length_rejected=1
oversized_build_response_error_code=7
result=PASS
~~~

The exact-limit input encrypted successfully and its Base64 ciphertext remained below the configured 1 MiB socket buffer. Both larger plaintext inputs were rejected before encryption. buildResponse() returned error code 7 ("Message encryption failed") for the oversized response. No ASan error occurred.

### Differential control against the old code

I compiled the **same caller** against pre-fix BrowserMessageBuilder.cpp at commit 26da2a1b. Under ASan the process aborted on the over-limit path:

~~~text
ERROR: AddressSanitizer: heap-buffer-overflow
READ of size 1
BrowserMessageBuilder::getQByteArray(unsigned char const*, unsigned int) const
0 bytes after a 1,048,576-byte allocation
~~~

This confirms that the caller reaches the affected code. The system Botan library in this Linux test was not ASan-instrumented, so ASan caught the caller's subsequent out-of-bounds **read** from the undersized ciphertext vector, not Botan's preceding out-of-bounds **write**. My earlier Windows debugger reproduction captured the crypto-path write access violation. I am not treating the Linux read report by itself as proof of an instrumented Botan write.

### Real keepassxc-proxy framing check

I built and ran the **actual merged keepassxc-proxy executable**, connected it to a mock KeePassXC local server under an isolated XDG_RUNTIME_DIR, and decoded the length-prefixed native-messaging frames emitted on its stdout. This uses the real proxy and local socket, but **does not run the browser extension**. For the fragmented case I waited for the first native frame before sending the second fragment, ensuring separate reads.

| Bytes sent over the local stream socket | Native frames emitted by the proxy | JSON parse |
| --- | --- | --- |
| {"action":"one","nonce":"abc"} | One complete frame | Valid |
| {"action":"one","nonce":"abcdef"} sent as two chunks, split between abc and def | Two frames containing the two partial chunks | Both invalid |
| {"action":"one"}{"action":"two"} sent together | One frame containing both objects concatenated | Invalid (extra data) |

Exact observed frames in the split case were '{"action":"one","nonce":"abc' and 'def"}'. The coalesced frame was '{"action":"one"}{"action":"two"}'. The complete-message control parsed correctly. The proxy did not crash and ASan reported no memory error.

This establishes that the **outgoing** local-socket-to-native-messaging code still frames each readAll() chunk as one JSON message. A stream-socket read is not guaranteed to align with a JSON object's boundaries. BrowserHost's new parser/tests cover the opposite, **incoming** direction and do not validate outgoing framing. These results are consistent with the invalid JSON observed in the earlier comment; however, I have not run the actual extension and cannot claim to have reproduced its UI error. I also cannot determine from these tests whether this behavior predates the PR.

### Assessment and remaining checks

The merged code rejects plaintext above 655,360 bytes before crypto_box_easy(), allocates space for the MAC, validates decoded key/nonce lengths, and returns an error for oversized browser responses. The boundary and differential tests support that the **reported memory overflow is fixed in the tested source build**; I found no bypass of that specific condition in these tests. The proxy's outgoing JSON framing is a separate functional issue to fix or track for the final release.

I have **not** tested a packaged Windows beta end-to-end with the browser extension. When a beta containing 80e68bb4 is available, I recommend rerunning the original authorized get-logins case (125 matching entries with 8,161-character titles): verify KeePassXC stays alive, returns a bounded/error response, and still serves a small authorized response. The extension/proxy path should also be checked with forced fragmentation and coalescence. I would not mark those packaged/extension checks as passed based on the source-level and mock-server results above.

**Comment by varjolintu** [comment]
> This establishes that the outgoing local-socket-to-native-messaging code still frames each readAll() chunk as one JSON message. A stream-socket read is not guaranteed to align with a JSON object's boundaries. BrowserHost's new parser/tests cover the opposite, incoming direction and do not validate outgoing framing.

The messages are already sent as under 1M packages from KeePassXC socket to the proxy socket after those are parsed in BrowserHost. If we want to make sure any fragmentation etc. does not happen between that transfer, the same check could be added to the proxy code as well.
